### PR TITLE
Implement QC Validation and Super-majority Stake Check

### PR DESCRIPTION
**Pull request by tholonious**

### CODE REVIEW

**Comment by tholonious** [comment]
Current dependencies on/for this PR:
* feature/proof-of-stake
  * **PR #649** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/649" target="_blank">Graphite</a> 
    * **PR #658** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/658" target="_blank">Graphite</a> 
      * **PR #664** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/664" target="_blank">Graphite</a> 
        * **PR #665** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/665" target="_blank">Graphite</a>  👈
        * **PR #663** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/663" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/deso-protocol/core/665?utm_source=stack-comment).

**Review by lazynina — APPROVED** [review]

File: consensus/utils.go [hunk 1]
```diff
@@ -6,9 +6,51 @@ import (
...
+
+	// Compute the total stake in the QC and the total stake in the network
+	stakeInQC := uint256.NewInt()
+	totalStake := uint256.NewInt()
+
+	// Fetch the validators in the QC
+	validatorPublicKeysInQC := []*bls.PublicKey{}
+
+	// Fetch the validators in the QC, and compute the sum of stake in the QC and in the network
+	for ii := range validators {
```
Comment (AeonSw4n) [inline consensus/utils.go:31]: Should we also validate that `len(validators) == qc.GetSignersList().Size()`?

File: consensus/utils.go [hunk 2]
```diff
@@ -154,3 +201,38 @@ func isValidSignature(publicKey *bls.PublicKey, signature *bls.Signature, payloa
 	isValid, err := bls.VerifyAggregateSignatureSinglePayload([]*bls.PublicKey{publicKey}, signature, payload)
 	return err == nil && isValid
 }
+
+// This function uses integer math to verify if the provided stake amount represents a
+// super-majority 2f+1 Byzantine Quorum. First we need the following context:
+// - Assume N = total stake in the network
+// - Assume f = safe faulty stake in the network
```
Comment (AeonSw4n) [inline consensus/utils.go:208]: I'm not sure what "safe faulty" means here. Should we just say "faulty stake" or "malicious stake"?

File: consensus/utils.go [hunk 3]
```diff
@@ -154,3 +201,38 @@ func isValidSignature(publicKey *bls.PublicKey, signature *bls.Signature, payloa
...
+// - Assume C = honest stake in the network
+//
+// - We define N = C + f, and N = 3f + 1
+// - We know that f = (N - 1) / 3
+// - We want to determine if we have a BQ where C >= 2f + 1
+//
+// Given the above, we can use derive the following condition as the super-majority QC check:
```
Comment (AeonSw4n) [inline consensus/utils.go:215]: Drop "use".

File: consensus/utils.go [hunk 4]
```diff
@@ -6,9 +6,51 @@ import (
...
+		totalStake.Add(totalStake, validators[ii].GetStakeAmount())
+	}
+
+	// Check if the QC contains a super-majority of stake
+	if !isSuperMajorityStake(stakeInQC, totalStake) {
+		return false
+	}
+
+	// Finally, validate the signature
+	isValidSignature, err := bls.VerifyAggregateSignatureSinglePayload(
```
Comment (AeonSw4n) [inline consensus/utils.go:45]: I'm getting a warning here that `isValidSignature` is already defined. Should we maybe rename it so it doesn't overlap with an existing method?

**Review by AeonSw4n — CHANGES_REQUESTED** [review]

File: consensus/utils.go (hunk 2 above)
Comment (tholonious) [reply consensus/utils.go:208]: Removed the "safe" so it now says "faulty stake"

File: consensus/utils.go (hunk 4 above)
Comment (tholonious) [reply consensus/utils.go:45]: Oof good catch. There was a function with the same name. Fixed now

File: consensus/utils.go (hunk 1 above)
Comment (tholonious) [reply consensus/utils.go:31]: No actually. The Bigint.ToBytes function compresses the byte encoding for a BigInt such if there's a sequence of 0 at higher indices, of the construct, then they are removed from the byte encoding and always assumed to be zero. This way, `len(validators) >= qc.GetSignersList().Size()` for Indices that exist in the validators slice but not in the signers list return a false value.

Ex: if we have a the sequence `true, true, true, false true, false, false, false, false, false`, the binary representation is `0000010111`, which gets compressed to `10111`.  Calling `bitset.Get(8)` returns a value of false for index 9.

**Review by AeonSw4n — APPROVED** [review]

**Pull request merged**